File: src/features/work-hours/domain/repositories/work-entry-repository.ts
import type {
  WorkDay,
  WorkDaySchedule,
  WorkEntry,
  WorkEntryInput,
} from "@/features/work-hours/domain/entities/work-hours";

export interface WorkEntryRepository {
  listByPresence(presenceId: string): Promise<WorkEntry[]>;
  /** Atomically replace the full entry set of a presence day. Returns the
   * canonical entries plus the schedule (its stored end recomputed). */
  replaceForPresence(
    presenceId: string,
    entries: WorkEntryInput[],
  ): Promise<WorkDay>;
  /** The day's start/end times, or `null` when never set. */
  getSchedule(presenceId: string): Promise<WorkDaySchedule | null>;
  /** Set the day's start time; the backend recomputes and stores the end. */
  setSchedule(
    presenceId: string,
    startMinutes: number,
  ): Promise<WorkDaySchedule>;
}

File: src-tauri/src/domain/repositories/work_entry_repository.rs
use async_trait::async_trait;

use crate::domain::entities::work_day_schedule::WorkDaySchedule;
use crate::domain::entities::work_entry::WorkEntry;
use crate::domain::error::DomainError;

/// Persistence contract for a day's work hours (entries + optional start/end
/// schedule), implemented over the unlocked vault connection. The day's entry
/// set is replaced as a whole: entry ids are not surfaced as stable handles.
#[async_trait]
pub trait WorkEntryRepository: Send + Sync {
    /// The entries recorded for a presence day, ordered by position.
    async fn list_by_presence(&self, presence_id: &str) -> Result<Vec<WorkEntry>, DomainError>;

    /// Atomically replace the full entry set of a presence day. Passing an
    /// empty slice clears the day.
    async fn replace_for_presence(
        &self,
        presence_id: &str,
        entries: &[WorkEntry],
    ) -> Result<(), DomainError>;

    /// The day's start/end times, if the user set them.
    async fn get_schedule(&self, presence_id: &str)
        -> Result<Option<WorkDaySchedule>, DomainError>;

    /// Create or update the day's start/end times.
    async fn set_schedule(&self, schedule: &WorkDaySchedule) -> Result<(), DomainError>;
}
